servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/publisher/TimeoutPublisherTest.java

```diff
@@ -348,6 +350,19 @@ public Cancellable execute(final Runnable task) throws RejectedExecutionExceptio
         assertThat(subscriber.awaitOnError(), instanceOf(TimeoutException.class));
     }
 
+    @Test
+    @RepeatedTest(100)
+    void timeoutDemandTimerCancellation() {
+        init(TimerBehaviorParam.DEMAND_TIMER, Duration.ofSeconds(1));
+        Subscription subscription = subscriber.awaitSubscription();
+        subscription.request(1);
+        assertThat(testExecutor.scheduledTasksPending(), is(0));
+        ForkJoinPool.commonPool().execute(() -> publisher.onNext(1));
+        assertThat(subscriber.takeOnNext(), is(1));
```

Comment:
This is a blocking call so we won't get concurrency between onNext and request. I think we can target the race w/out looping (test would be racy but if it every fails we have a problem)

```
Executor executor = ...;
CountDownLatch aLatch = ...;
// initial demand == 1

Thread 1 -> onNext(1) -> `startTimer`, `executor.schedule(..)` does `aLatch.countDown(); Thread.sleep(1000);`
Thread 2 -> aLatch.await(), request(1) -> `stopTimer`
Thread 1 -> sets `timerCancellable`
Thread 2 -> unsets `timerCancellable`, and cancels()

// we should be able to wrap TestExecutor and advance the time here
assertThat(subscriber.pollTerminal(<timeout interval * 2>), is(nullValue()));
```

Reply:
If I comment out the new `if` branch in the fix the test fails nearly 10% of the time on my local machine so it definitely stresses the race.

Wrt `.takeOnNext()` you're right: it is blocking, and waits on the result of the `publisher.onNext(1)` executing in `ForkJoinPool`. As it turns out, that is helpful as it causes the call to `target.onNext(..)` to potentially trigger context switches which immediately proceeds the window for the race. If we omit the `.takeOnNext()` call it requires on the order of 10k iterations to trigger the bug (on my machine).

I think you're right that we can engineer an `Executor` that pauses in the `.schedule(..)` call, but is it worth our time if we can just repeat the test 100 times?

Reply:
In meeting so I may have missed the sequencing, but if your current test does exercise the race than feel free to 🚢